src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1458,17 +1458,32 @@ internal static bool IsCustomTaskType(this NamedTypeSymbol type, out object buil
             var arity = type.Arity;
             if (arity < 2)
             {
-                // Find the AsyncBuilder attribute.
-                foreach (var attr in type.GetAttributes())
+                // special case for SourceNamedTypeSymbol which early decodes method builder to prevent possible binding loops
+                if (type is SourceNamedTypeSymbol namedTypeSymbol)
```

Comment:
>if (type is SourceNamedTypeSymbol namedTypeSymbol) [](start = 16, length = 50)

If we go with this approach (of using an early decoded attribute), we should hide the details behind an `IsCustomTaskType(out object builderArgument)` virtual method on `NamedTypeSymbol`, similar to other such attributes (e.g.: `HasCodeAnalysisEmbeddedAttribute`).

Reply:
Moved from extension to virtual methods.